ledger/state/elite_edge_node_pool.go

```diff
@@ -0,0 +1,279 @@
+package state
+
+import (
+	crand "crypto/rand"
+	"fmt"
+	"io"
+	"log"
+	"math/big"
+
+	"github.com/thetatoken/theta/common"
+	"github.com/thetatoken/theta/common/util"
+	"github.com/thetatoken/theta/core"
+	"github.com/thetatoken/theta/crypto/bls"
+	"github.com/thetatoken/theta/ledger/types"
+)
+
+const eenpRewardN = 400 // Reward receiver sampling params
+
+//
+// ------- EliteEdgeNodePool ------- //
+//
+
+type EliteEdgeNodePool struct {
+	readOnly bool
+	sv       *StoreView
+}
+
+// NewEliteEdgeNodePool creates a new instance of EliteEdgeNodePool.
+func NewEliteEdgeNodePool(sv *StoreView, readOnly bool) *EliteEdgeNodePool {
+	return &EliteEdgeNodePool{
+		readOnly: readOnly,
+		sv:       sv,
+	}
+}
+
+// Contains checks if given address is in the pool.
+func (eenp *EliteEdgeNodePool) RandomRewardWeight(block common.Hash, eenAddr common.Address) int {
+	een := eenp.Get(eenAddr)
+	if een == nil {
+		return 0
+	}
+	totalStake := eenp.sv.GetTotalEENStake()
+	stake := een.TotalStake()
+
+	return sampleEENWeight(util.NewHashRand(block.Bytes()), stake, totalStake)
+}
+
+func sampleEENWeight(reader io.Reader, stake *big.Int, totalStake *big.Int) int {
+	b := new(big.Int).Div(stake, core.MinEliteEdgeNodeStakeDeposit)
+
+	base := new(big.Int).SetUint64(1e18)
+
+	p := new(big.Int).Mul(base, big.NewInt(eenpRewardN))
+	p.Mul(p, stake)
+	p.Div(p, totalStake)
```

Comment:
Need to add this between Line 55 and Line 56?

p.Mul(p, stake)
p.Div(core.MinEliteEdgeNodeStakeDeposit)

So that p = n * S_min / S_total * a, where a = (S/S_min) / floor(S/S_min)